---
title: "Tissue Prediction"
author: "Shannon E. Ellis"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{recount quick start guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()
```
# Load libraries

```{r load-packages, message = FALSE, warning = FALSE}
## load libraries
library(devtools)
document("/users/sellis/phenopredict")
library('phenopredict')
library(rtracklayer)
library(recount)

## Set colors
## import colors to use
  bright= c(red=rgb(222,45,38, maxColorValue=255), #de2d26
            pink=rgb( 255, 102, 153, maxColorValue=255), #ff6699
            orange=rgb(232,121,12, maxColorValue=255),   #e8790c
            yellow=rgb(255,222,13, maxColorValue=255), #ffde0d          
            green=rgb(12,189,24, maxColorValue=255),  #0cbd18           
            teal=rgb(59,196,199, maxColorValue=255), #3bc4c7
            blue=rgb(58,158,234, maxColorValue=255), #3a9eea
            purple=rgb(148,12,232, maxColorValue=255)) #940ce8  
```

# Load data and select regions
```{r 'load-data-select-regions', message = FALSE, warning = FALSE}

load("/dcl01/leek/data/sellis/barcoding/data/sample_individuals.Rdata")
load("/dcl01/leek/data/sellis/barcoding/data/rda/gtexmetadata.rda")
gtex_meta = gtexmetadata
gtex_meta = cbind(gtex_meta,usegtex)
rm(gtexmetadata,usegtex)
gtex_meta[is.na(gtex_meta[,"SMTS"]),"SMTS"] <- c("Esophagus","Skin","Stomach","Skin","Esophagus")
usegtex = gtex_meta$usegtex
pheno = gtex_meta
pheno = pheno[usegtex,]
pheno = pheno[sample_individuals,]

# prepare expression data
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/merge_input_SMTS_nocovars.rda")){
	source("/dcl01/leek/data/sellis/barcoding/scripts/gtexLoad.R")
	chrs = c("X","Y",1:22)
	for(i in 1:length(chrs)){
	  chrname=paste0("chr",chrs[i])
	  cat(paste0(chrname,"..."))
	  dat = gtexLoad(chrname)
	  cm = dat$coverageMatrix
	  regiondata = dat$regions
	  cm = log2(cm + 1)
	  wid = width(dat$regions)

	  cm = cm[, usegtex]
	  cm_all=cm[wid>20,]
	  regiondata = regiondata[wid>20]
	  cm_regiondata = regiondata
	  #only include the half of samples used for building the predictor
	  cm = cm_all[,sample_individuals]
	  cm_new = cm_all[,-sample_individuals]
	  assign(paste0("cm_new",'_',chrname),cm_new)
	  assign(paste0("regiondata","_",chrname),cm_regiondata)
	# number of regions in expression data 
	nrow(cm)

	# number of samples included in example
	ncol(cm)
	inputdata<-select_regions(expression=cm, regiondata=regiondata ,phenodata=pheno, phenotype="SMTS",
	    covariates=NULL,type="factor", numRegions=100)
	assign(paste0("inputdata",'_',chrname),inputdata)
  }
    save(cm_new_chr1, cm_new_chr2, cm_new_chr3, cm_new_chr4, cm_new_chr5, cm_new_chr6, cm_new_chr7, cm_new_chr8, cm_new_chr9, cm_new_chr10, cm_new_chr11, cm_new_chr12, cm_new_chr13, cm_new_chr14, cm_new_chr15, cm_new_chr16, cm_new_chr17, cm_new_chr18, cm_new_chr19, cm_new_chr20, cm_new_chr21, cm_new_chr22, cm_new_chrX, cm_new_chrY, file="data/cm_new_SMTS_nocovars.Rda")
  save(regiondata_chr1, regiondata_chr2, regiondata_chr3, regiondata_chr4, regiondata_chr5, regiondata_chr6, regiondata_chr7, regiondata_chr8, regiondata_chr9, regiondata_chr10, regiondata_chr11, regiondata_chr12, regiondata_chr13, regiondata_chr14, regiondata_chr15, regiondata_chr16, regiondata_chr17, regiondata_chr18, regiondata_chr19, regiondata_chr20, regiondata_chr21, regiondata_chr22, regiondata_chrX, regiondata_chrY, file="data/cm_regiondata_SMTS_nocovars.Rda")
}

# c("Sex","AGE","BMI","SMNABTCHT")
```

# Merge input data

```{r 'merge-inputdata', message = FALSE, warning = FALSE}
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/merge_input_SMTS_nocovars.rda")) {
	# test = as.list(file.list)
	file.list.inputdata <- grep("inputdata_chr",ls(),val=T)
	# test<- merge_input(inputdata_list=list(do.call(get(file.list.inputdata))))

	merge_input<- merge_input(inputdata_list=list(inputdata_chr1, inputdata_chr2, inputdata_chr3, inputdata_chr4, inputdata_chr5, inputdata_chr6, inputdata_chr7, inputdata_chr8, inputdata_chr9, inputdata_chr10, inputdata_chr11, inputdata_chr12, inputdata_chr13, inputdata_chr14, inputdata_chr15, inputdata_chr16, inputdata_chr17, inputdata_chr18, inputdata_chr19, inputdata_chr20, inputdata_chr21, inputdata_chr22, inputdata_chrX, inputdata_chrY) )
	save(merge_input,file="/dcl01/leek/data/sellis/barcoding/data/merge_input_SMTS_nocovars.rda")
} else {
	load("/dcl01/leek/data/sellis/barcoding/data/merge_input_SMTS_nocovars.rda")
}
# taking a look at output of select(regions)
dim(merge_input$covmat)
merge_input$regiondata
```

# Build predictor

```{r 'build-predictor', message = FALSE, warning = FALSE}
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/merge_input_SMTS_nocovars.rda")) {
	predictor<-build_predictor(inputdata=merge_input ,phenodata=pheno, phenotype="SMTS",		 
    	covariates=NULL,type="factor", numRegions=40)
	save(predictor, file= "/dcl01/leek/data/sellis/barcoding/data/build_predictor_SMTS_nocovars.rda")
}else{
	load('/dcl01/leek/data/sellis/barcoding/data/build_predictor_SMTS_nocovars.rda')
}

#number of probes used for prediction
length(predictor$trainingProbes)

#this contains the coefficient estimates used for prediction. 
# the number of rows corresponds to the number of sites used for prediction
# while the columns corresponds to the number of categories of your phenotype.
dim(predictor$coefEsts)

```
# Test predictor

```{r 'test-predictor', message = FALSE, warning = FALSE}

predictions_test <-test_predictor(inputdata=merge_input ,phenodata=pheno, phenotype="SMTS",covariates=NULL,type="factor",predictordata=predictor )
# number of samples
length(predictions_test$predicted)

# get summary of how prediction is doing
predictions_test$summarized

actual = predictions_test$actual
predicted = predictions_test$predicted
perc_correct = predictions_test$summarized[,"percent_correct"]

#get output for package
predictions_gtex = as.data.frame(cbind(pheno$Run,"gtex",actual=as.character(actual),predicted,perc_correct))
```

# Extract data

```{r 'extract-data', message = FALSE, warning = FALSE}
# looking at the input data for extract_data
# file.list.cm <- grep("cm_new_chr",ls(),val=T)
# file.list.inputdata <- grep("inputdata_chr",ls(),val=T)
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/merge_test_SMTS_nocovars.rda")) {
	for(i in 1:length(chrs)){
		chr=chrs[i]
		input = paste0("cm_new_chr",chr)
		input2 = paste0("regiondata_chr",chr)
    	new<-extract_data(newexpression=get(input), newregiondata=get(input2), predictordata=predictor)
    	assign(paste0("new_chr",chr),new)
    }


	file.list.inputdata <- ls()[startsWith(ls(),"new_chr")]
	merge_test<- merge_input(inputdata_list=list(new_chr1, new_chr2, new_chr3, new_chr4, new_chr5, new_chr6, new_chr7, new_chr8, new_chr9, new_chr10, new_chr11, new_chr12, new_chr13, new_chr14, new_chr15, new_chr16, new_chr17, new_chr18, new_chr19, new_chr20, new_chr21, new_chr22, new_chrX, new_chrY))

  ov <- findOverlaps(predictor$regiondata,merge_test$regiondata)
  index_regions <- subjectHits(ov)
  #reorder GRanges objects
  merge_test$regiondata <- merge_test$regiondata[index_regions]
  merge_test$covmat <- merge_test$covmat[index_regions,]
  save(merge_test, file="/dcl01/leek/data/sellis/barcoding/data/merge_test_SMTS_nocovars.rda")
}else{
	load("/dcl01/leek/data/sellis/barcoding/data/merge_test_SMTS_nocovars.rda")
}
```

# Predict Tissue in GTEx

```{r 'predict-phenotype', message = FALSE, warning = FALSE}
predictions<-predict_pheno(inputdata_test= merge_test, phenodata=pheno, phenotype="SMTS", covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions)

#since we know the truth here, let's check and see how we're doing:
 
 #pull out actual Sex
  pheno2 = gtex_meta
  pheno2 = pheno2[usegtex,]
  pheno2 = pheno2[-sample_individuals,]
  actual = pheno2$SMTS

  #define predicted
  predicted = predictions

  #summarize data
	number_match <- sum(predicted==actual)
	perc_correct = sum(predicted==actual)/length(actual)
	number_sites = nrow(predictor$coefEsts)

	summarized = c(number_sites,number_match, perc_correct)
	names(summarized) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 summarized

 #get output for package
predictions_gtex_test = as.data.frame(cbind(pheno2$Run,"gtex",actual=as.character(actual),predicted,perc_correct))
predictions_gtex$perc_correct <- perc_correct

```

# Predict Tissue in TCGA

```{r 'TCGA', message = FALSE, warning = FALSE}

#extract names of possible tissues predicted
tiss = names(table(actual))
tissNA = c(tiss, NA)


## save as bedfile for region extraction in TCGA
#compare to previous sex prediction
# gr = regions
# export.bed(gr, con="/dcl01/leek/data/sellis/barcoding/TCGA/TCGA_Tissue_v2.bed")

# get phenotype info in order
recount::all_metadata('TCGA') -> md 
bws = gsub('.bw','', md$bigwig_file)
md2 = md[!is.na(md$xml_tumor_tissue_site),]

if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/TCGA_Tissue_inputdata.rda")) {

	## Extract same regions from TCGA data
	regions = predictor$regiondata
	load('/dcl01/lieber/ajaffe/lab/gtex_ers/rse_TCGA.Rdata')
	findOverlaps(regions,rse) -> ov
	## extract regions from predictor
	rse[subjectHits(ov)] -> rse_sub
	## get coverage matrix ; order same as in regions
	## note: these aren't scaled
	cov_tcga <- assays(rse_sub)$counts




	#ensure covmat is in same order as regions
	if( identical(tolower(bws),colnames(cov_tcga) )==TRUE){
		covmat= cov_tcga
		regions = regions

		#remove coverage samples for which we don't have metadata
		covmat = covmat[,colnames(covmat) %in% tolower(bws)]

		#put data on a log2 scale
		dm = log2(covmat[,!is.na(md$xml_tumor_tissue_site)]+1)
	}
	## only include samples with tissues included in GTEx
	### Keep only the good SRA
	tiss_in_gtex = md2$xml_tumor_tissue_site %in% tiss
	tcgaGene = dm[,tiss_in_gtex]
	tcga_meta2 = md2[tiss_in_gtex,]


	TCGA_inputdata=c()
	TCGA_inputdata$covmat <- tcgaGene
	TCGA_inputdata$regiondata <- predictor$regiondata

	  ov <- findOverlaps(predictor$regiondata,TCGA_inputdata$regiondata)
	  index_regions <- subjectHits(ov)
	  #reorder GRanges objects
	  TCGA_inputdata$regiondata <- TCGA_inputdata$regiondata[index_regions]
	  TCGA_inputdata$covmat <- TCGA_inputdata$covmat[index_regions,]

	  save(TCGA_inputdata, tcga_meta2, file="/dcl01/leek/data/sellis/barcoding/data/TCGA_Tissue_inputdata.rda")
}else{
	load("/dcl01/leek/data/sellis/barcoding/data/TCGA_Tissue_inputdata.rda")
}

predictions<-predict_pheno(inputdata_test= TCGA_inputdata, phenodata=pheno, phenotype="SMTS", covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions)

actual = as.character(tcga_meta2$xml_tumor_tissue_site)
#define predicted
predicted = predictions

#summarize data
number_match <- sum(predicted==actual)
perc_correct = sum(predicted==actual)/length(actual)
number_sites = nrow(predictor$coefEsts)

TCGA_summarized = c(number_sites,number_match, perc_correct)
names(TCGA_summarized) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 TCGA_summarized

#get output for package
predictions_tcga = as.data.frame(cbind(tcga_meta2$gdc_file_id,"tcga",actual,predicted,perc_correct))

## How do we do in healthy controls
ht <- tcga_meta2$cgc_sample_sample_type=="Solid Tissue Normal" & !(is.na(tcga_meta2$cgc_sample_sample_type))
cancer <- tcga_meta2$cgc_sample_sample_type!="Solid Tissue Normal" & !(is.na(tcga_meta2$cgc_sample_sample_type))


predicted_cancer <- predicted[cancer]
actual_cancer <- actual[cancer]

toremove = !is.na(predicted_cancer)
predicted_cancer <- predicted_cancer[toremove]
actual_cancer <- actual_cancer[toremove]
length(actual_cancer)

perc_correct_cancer <- sum(predicted_cancer==actual_cancer)/length(actual_cancer)
perc_correct_cancer

predicted_ht <- predicted[ht]
actual_ht <- actual[ht]

toremove = !is.na(predicted_ht)
predicted_ht <- predicted_ht[toremove]
actual_ht <- actual_ht[toremove]
length(actual_ht)

perc_correct_ht <- sum(predicted_ht==actual_ht)/length(actual_ht)
perc_correct_ht

```

# Predict Tissue in SRA

```{r 'SRA', message = FALSE, warning = FALSE}
## load SRA metadata

## extract regions from SRA data
if(!file.exists('/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_Tissue.Rdata')) {
  #### SRA data [generated in merge_bwtool_by_chunk.R]
  ## which chunk regions are in [chunk_grp]
  load('/dcl01/leek/data/sellis/barcoding/data/chunk_grp.Rdata')
  ## region information [regions]
  load('/dcl01/leek/data/sellis/barcoding/data/regions-cut0.5.Rdata')
  reg = regions
  rm(regions)
  ## add chunk information to all 1,187,643 regions
  reg$chunkname <- chunk_grp

  ## GTEx regions to subset
  regions_to_subset = predictor$regiondata

  ## get regions overlap [b/w all quantified regions and GTEx subset]
  ov <- findOverlaps(reg,regions_to_subset)
  ## Integer vector to re-order regions to https://github.com/jtleek/sra-gtex/blob/master/select-gtex-ers.R#L80-L88
  index_regions <- subjectHits(ov)
  index_regions_query <- queryHits(ov)
  #reorder GRanges objects
  reg <- reg[index_regions_query,]
  regions_to_subset <- regions_to_subset[index_regions,]

  regions_split <- split(reg, reg$chunkname)


  ## load the coverageMatrix of interest from the appropriate "chunk"

  # dir.create('data/coverageMatrix_subset', showWarnings = FALSE)

  xx <- mapply(function(regs, chunk_name) {
      message(paste(Sys.time(), 'processing', chunk_name))
      
      ## Load coverage matrix for each chunk [coverageMatrix]
      load(paste0('/dcl01/leek/data/gtex_work/runs/sra/DER_analysis/coverageMatrix/ers_gtex/coverageMatrix-cut0.5-', chunk_name, '.Rdata'))

      ## Load regions included in the chunk [regions_subset]
     load( paste0('/dcl01/leek/data/gtex_work/runs/sra/DER_analysis/coverageMatrix/ers_gtex/regions_', chunk_name, '-cut0.5.Rdata'))

      ## Identify which regions to keep
      regions_keep <- which(countOverlaps( regions_subset,regions_to_subset) > 0)
      stopifnot(length(regions_keep) == length(regs))
      
      ## Save actual subset of regions used
      regions_subset <- regions_subset[regions_keep]
      save(regions_subset, file = paste0('/dcl01/leek/data/sellis/barcoding/data/minfi_rda/regions_', chunk_name, '-cut0.5.Rdata'))
      
      ## Compute chunk coverage matrix
      coverageMatrix_subset <- coverageMatrix[regions_keep,]
      save(coverageMatrix_subset, file = paste0('/dcl01/leek/data/sellis/barcoding/data/minfi_rda/coverageMatrix-cut0.5-', chunk_name,
          '.Rdata'))
      
      ## Finish
      return(invisible(NULL))
  }, regions_split, names(table(reg$chunkname)))

  ## compile the subset of regions into a single object
  chunks <-names(table(reg$chunkname))
  coverageMatrix=c()
  regions = GRanges()

  for (i in 1:length(chunks)){
        message(paste(Sys.time(), 'processing', chunks[i]))

    ## load in each chunk and save to a data frame
    load(paste0('/dcl01/leek/data/sellis/barcoding/data/minfi_rda/coverageMatrix-cut0.5-', chunks[i],
          '.Rdata'))
    coverageMatrix = rbind(coverageMatrix, coverageMatrix_subset)

    ## load in and combine region data
    load(paste0('/dcl01/leek/data/sellis/barcoding/data/minfi_rda/regions_', chunks[i], '-cut0.5.Rdata'))
    regions = append(regions,regions_subset)
  } 
  	
  	sra_inputdata=list()
	sra_inputdata$covmat <- coverageMatrix
	sra_inputdata$regiondata <- regions

  save(sra_inputdata,file='/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_Tissue.Rdata')
}else{
  	load('/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_Tissue.Rdata')
}



## load SRA metadata
### Load in SRA metadata
load('/dcl01/leek/data/recount-website/metadata/metadata_sra.Rdata')
metadata <- metadata[!is.na(metadata$bigwig_path), ]
sra_meta = metadata
rm(metadata)


### Keep only the good SRA and get them in the same order
SRA_inputdata = sra_inputdata

mm = match(colnames(SRA_inputdata$covmat),sra_meta$run)
sra_meta = sra_meta[mm,]
pd = read_csv("https://raw.githubusercontent.com/nellore/runs/master/sra/v2/hg38/SraRunInfo.csv")
sra_meta = left_join(as.data.frame(sra_meta),pd,by=c("run"="Run","sample"="Sample"))

## make sure data are in the same order as predictor data  
ov <- findOverlaps(predictor$regiondata,SRA_inputdata$regiondata)
index_regions <- subjectHits(ov)
#reorder GRanges objects
SRA_inputdata$regiondata <- SRA_inputdata$regiondata[index_regions]
SRA_inputdata$covmat <- SRA_inputdata$covmat[index_regions,]
#log2 scale
SRA_inputdata$covmat =  log2(SRA_inputdata$covmat +1)

## make predictions
predictions<-predict_pheno(inputdata_test= SRA_inputdata, phenodata=pheno, phenotype="SMTS", covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions)

## test for accuracy
actual_all = sra_meta$sharq_beta_tissue
#define predicted
predicted_all = predictions

## only include those in accuracy assesment for which we have predictions from sharq (NLP)
keep = !is.na(sra_meta$sharq_beta_tissue) & sra_meta$sharq_beta_tissue %in% tolower(tiss)
actual = actual_all[keep]
predicted = predicted_all[keep]
predicted = tolower(predicted)
length(predicted)

#summarize data
number_match <- sum(predicted==actual)
perc_correct = sum(predicted==actual)/length(actual)
number_sites = nrow(predictor$coefEsts)

SRA_summarized = c(number_sites,number_match, perc_correct)
names(SRA_summarized) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 SRA_summarized

 #get output for package
predictions_sra = as.data.frame(cbind(sra_meta$run,"sra",actual=as.character(actual_all),predicted=predicted_all,perc_correct))

```

```{r 'plot-output', message = FALSE, warning = FALSE, fig.width=8}

data = rbind(predictions_test$summarized,summarized,TCGA_summarized,SRA_summarized) %>% data.frame()
t(data) %>% as.matrix() -> mat
colnames(mat) <- c("GTEx: training","GTEx: test","TCGA", "SRA")
mat 

colours2 = c(bright["purple"],bright["purple"],bright["pink"],bright["teal"])

# library(extrafont)
# pdf("/dcl01/leek/data/sellis/barcoding/plots/TissuePrediction.pdf",family="Roboto Condensed",width=12, height=6)
par(cex=1.3)
barplot(as.numeric(mat["percent_correct",]*100), names.arg=c("GTEx: training","GTEx: test","TCGA", "SRA"),
        main="Tissue", ylim=c(0,100), ylab="Accuracy", xlab="Data Set Used", col=colours2, cex.lab=1.3, cex.axis=1.2, cex.main=1.4)
axis(1, at=c(0,5), xpd=T, lwd.ticks = 0, labels=F)
#dev.off()


## add in tcga_ht and tcga_cancer

toplot = as.numeric(mat["percent_correct",])*100
names(toplot) <- c("GTEx: training","GTEx: test","TCGA", "SRA")
toplot["TCGA_ht"] <- perc_correct_ht*100
toplot["TCGA_cancer"] <- perc_correct_cancer*100
toplot = toplot[-3]
toplot = toplot[c(1,2,4,5,3)]
colours2 = c(bright["purple"],bright["purple"],bright["pink"],bright["pink"],bright["teal"])


# pdf("/dcl01/leek/data/sellis/barcoding/plots/TissuePrediction_ht_cancer.pdf",family="Roboto Condensed",width=12, height=6)
par(cex=1.3)
barplot(as.numeric(toplot), names.arg=c("GTEx: training","GTEx: test","TCGA: healthy tissue","TCGA: cancer", "SRA"),
        main="Tissue Prediction", ylim=c(0,100), ylab="Accuracy", xlab="Data Set Used", col=colours2, cex.lab=1.3, cex.axis=1.2, cex.main=1.4)
axis(1, at=c(0,5), xpd=T, lwd.ticks = 0, labels=F)
# dev.off()

```

```{r 'pheno-output', message = FALSE, warning = FALSE}
predictions_Tissue <- rbind(predictions_gtex,predictions_gtex_test,predictions_tcga,predictions_sra)
colnames(predictions_Tissue) <- c("sample_id","dataset","reported_tissue","predicted_tissue","accuracy_tissue") 
predictions_Tissue$dataset <- factor(predictions_Tissue$dataset)
predictions_Tissue$reported_tissue <- factor(predictions_Tissue$reported_tissue)
predictions_Tissue$predicted_tissue <- factor(predictions_Tissue$predicted_tissue)
predictions_Tissue$accuracy_tissue <- as.numeric(predictions_Tissue$accuracy_tissue)

save(predictions_Tissue, file="/dcl01/leek/data/sellis/barcoding/output/predictions_Tissue.rda")
```

```{r 'region-information', message = FALSE, warning = FALSE}
## Get required information for the plots
if(!file.exists('/dcl01/leek/data/sellis/barcoding/output/Tissue_regions.rda')) {	
	txdb <- GenomicFeatures::makeTxDbFromGFF('ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/gencode.v25.annotation.gff3.gz', format = 'gff3', organism = 'Homo sapiens')
	tx<-annotateTranscripts(txdb, annotationPackage = NULL, by = c("tx","gene"), codingOnly=FALSE, verbose = TRUE, requireAnnotation = FALSE)
	output<-matchGenes(predictor$regiondata, tx, type = c("any", "fiveprime"), promoterDist = 2500, skipExons = FALSE, verbose = TRUE)
	save(output, file= "/dcl01/leek/data/sellis/barcoding/output/Tissue_regions.rda")
}else{
	load('/dcl01/leek/data/sellis/barcoding/output/Tissue_regions.rda')
	table(output$description)
}
```


# Vignette information

```{r reproducibility}
## Time spent creating this report:
diff(c(startTime, Sys.time()))

## Date this report was generated
message(Sys.time())

## Reproducibility info
options(width = 120)
devtools::session_info()
```

Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library('rmarkdown')
system.time(render('/dcl01/leek/data/sellis/barcoding/phenotype_vignettes/phenopredict-Tissue.Rmd', 'BiocStyle::html_document'))

## Extract the R code
library('knitr')
knit('/dcl01/leek/data/sellis/barcoding/phenotype_vignettes/phenopredict_predictions/phenopredict-Tissue.Rmd', tangle = TRUE)
```


