---
title: "Sequencing Strategy Prediction"
author: "Shannon E. Ellis"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{recount quick start guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()
```
# Load libraries

```{r load-packages, message = FALSE, warning = FALSE}
## load libraries
library(devtools)
document("/users/sellis/phenopredict")
library('phenopredict')
library(rtracklayer)
library(recount)
library(genefilter)
library(plyr)
library(GenomicRanges)


## Set colors
## import colors to use
  bright= c(red=rgb(222,45,38, maxColorValue=255), #de2d26
            pink=rgb( 255, 102, 153, maxColorValue=255), #ff6699
            orange=rgb(232,121,12, maxColorValue=255),   #e8790c
            yellow=rgb(255,222,13, maxColorValue=255), #ffde0d          
            green=rgb(12,189,24, maxColorValue=255),  #0cbd18           
            teal=rgb(59,196,199, maxColorValue=255), #3bc4c7
            blue=rgb(58,158,234, maxColorValue=255), #3a9eea
            purple=rgb(148,12,232, maxColorValue=255)) #940ce8  
```

# Load SRA metadata
```{r 'load-sra-metadata', message = FALSE, warning = FALSE}

## load SRA metadata
### Load in SRA metadata
   #### SRA data [generated in merge_bwtool_by_chunk.R]
  ## which chunk regions are in [chunk_grp]
  load('/dcl01/leek/data/sellis/barcoding/data/chunk_grp.Rdata')
  ## region information [regions]
  load('/dcl01/leek/data/sellis/barcoding/data/regions-cut0.5.Rdata')
  reg = regions
  rm(regions)
  ## add chunk information to all 1,187,643 regions
  reg$chunkname <- chunk_grp

    ## have to go through and select regions from each chunk that specify library prep
	regions_split <- split(reg, reg$chunkname)
	phenot="LibraryLayout"

  ## remove noncanoncial chromosomes
  chunk_name = table(chunk_grp) %>% names()
  chunk_name = grep("chrUn", chunk_name,invert=T,val=T)
  chunk_name = grep("random", chunk_name,invert=T,val=T)
  chunk_name = grep("EBV", chunk_name,invert=T,val=T)
  chunk_name = grep("chrM", chunk_name,invert=T,val=T)

 # selecting training and test set samples
 load(paste0('/dcl01/leek/data/gtex_work/runs/sra/DER_analysis/coverageMatrix/ers_gtex/coverageMatrix-cut0.5-', chunk_name[1], '.Rdata'))
  set.seed(1567)
  samples_to_use<-sample(colnames(coverageMatrix), ncol(coverageMatrix)/2 )
  keep = colnames(coverageMatrix) %in% samples_to_use

load('/dcl01/leek/data/recount-website/metadata/metadata_sra.Rdata')
metadata <- metadata[!is.na(metadata$bigwig_path), ]
sra_meta = metadata
rm(metadata)


### Keep only the good SRA and get them in the same order
mm = match(colnames(coverageMatrix),sra_meta$run)
sra_meta = sra_meta[mm,]
pd = read_csv("https://raw.githubusercontent.com/nellore/runs/master/sra/v2/hg38/SraRunInfo.csv")
sra_meta = left_join(as.data.frame(sra_meta),pd,by=c("run"="Run","sample"="Sample"))

 sra_meta_training <<- sra_meta[keep,]
 sra_meta_test <<- sra_meta[!keep,]
```

# Select regions from training data

```{r 'select-regions', message = FALSE, warning = FALSE}
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_SequencingStrategy.Rdata")){

	xx <- lapply(chunk_name, function(chunk_name) {
	     message(paste(Sys.time(), 'processing', chunk_name))
	      
	     ## Load coverage matrix for each chunk [coverageMatrix]
	     load(paste0('/dcl01/leek/data/gtex_work/runs/sra/DER_analysis/coverageMatrix/ers_gtex/coverageMatrix-cut0.5-', chunk_name, '.Rdata'))

	      ## Load regions included in the chunk [regions_subset]
	     load(paste0('/dcl01/leek/data/gtex_work/runs/sra/DER_analysis/coverageMatrix/ers_gtex/regions_', chunk_name, '-cut0.5.Rdata'))

	     ## split into training and test 
	     sra_training = coverageMatrix[,keep]
	     sra_test = coverageMatrix[,!keep]



	     ## quick and dirty test to minimize number of regions input 
	     if(length(names(table(sra_meta_training[,phenot])))==2){
	     	subs = rowttests(sra_training,factor(sra_meta_training[,phenot]))
	     }else(
	     	subs = rowftests(sra_training,factor(sra_meta_training[,phenot]))
	     )

	     cutoff = quantile(subs$p.value,0.05)
	     tokeep<-subs$p.value<cutoff
	    sra_training = log2(sra_training[tokeep,]+1)
	    sra_test = log2(sra_test[tokeep,]+1)
		regions_subset = regions_subset[tokeep]

	      ## Select regions associated with LibraryPrep
   		 inputdata<-select_regions(expression=sra_training, regiondata=regions_subset ,phenodata=sra_meta_training, phenotype=phenot, covariates=NULL,type="factor", numRegions=10)

   		new<-extract_data(newexpression=sra_test, newregiondata=regions_subset, predictordata=inputdata)
    	inputdata$sra_test <- new
	     
	      ## Finish
	      return(inputdata)
	  })
	names(xx) <- chunk_name

	  ## compile the subset of regions into a single object
	  merge_input<- merge_input(inputdata_list=xx)

	#remove "id"column
	merge_input$covmat = dplyr::select(merge_input$covmat, -(.id))

	  save(merge_input,file='/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_SequencingStrategy.Rdata')
	}else{
	  load('/dcl01/leek/data/sellis/barcoding/data/SRA_inputdata_SequencingStrategy.Rdata')
	}

# taking a look at output of select_regions)
dim(merge_input$covmat)
merge_input$regiondata

```


# Build predictor in SRA training data

```{r 'build-predictor', message = FALSE, warning = FALSE}
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/build_predictor_SequencingStrategy_nocovars.rda")) {
	predictor<-build_predictor(inputdata=merge_input ,phenodata=sra_meta_training, phenotype=phenot,		 
		covariates=NULL,type="factor", numRegions=40)
	save(predictor, file= "/dcl01/leek/data/sellis/barcoding/data/build_predictor_SequencingStrategy_nocovars.rda")
}else{
	load('/dcl01/leek/data/sellis/barcoding/data/build_predictor_SequencingStrategy_nocovars.rda')
}

#number of probes used for prediction
length(predictor$trainingProbes)

#this contains the coefficient estimates used for prediction. 
# the number of rows corresponds to the number of sites used for prediction
# while the columns corresponds to the number of categories of your phenotype.
dim(predictor$coefEsts)
```

# Resubstitution Error (in SRA training data)

```{r 'test-predictor', message = FALSE, warning = FALSE}
predictions_test <-test_predictor(inputdata=merge_input ,phenodata=sra_meta_training, phenotype=phenot, 
    covariates=NULL,type="factor",predictordata=predictor )
# number of samples
length(predictions_test$predicted)

# get summary of how prediction is doing
predictions_test$summarized

summarized_training = predictions_test$summarized

predicted = predictions_test$predicted
actual = sra_meta_training[,phenot]

perc_correct = as.numeric(sum(predicted==actual)/length(actual))

 #get output for package
predictions_SequencingStrategy = data.frame(cbind(sra_meta_training$run,"sra",actual=as.character(actual),predicted=predicted,perc_correct))


## how do predictions fare when sra_misreported_paired_end==TRUE
set <- actual!=predicted
mismatch = sra_meta_training$sra_misreported_paired_end[set]
table(mismatch)
#Proportion of samples with misreported paired end sequencing that are 'incorrect' predictions in SRA training data
length(mismatch[mismatch==TRUE])/length(sra_meta_training$sra_misreported_paired_end[sra_meta_training$sra_misreported_paired_end==TRUE])
```

# Extract data for SRA test set

```{r 'extract-data', message = FALSE, warning = FALSE}
# looking at the input data for extract_data
if(!file.exists("/dcl01/leek/data/sellis/barcoding/data/sra_test_inputdata_SequencingStrategy_nocovars.rda")) {

	purrr::map(xx, function(x){return(x$sra_test)}) %>% ldply(., data.frame) -> sra_test_covmat
	#remove "id"column
	sra_test_covmat= dplyr::select(sra_test_covmat, -(.id))
	colremove<-grep("regiondata", colnames(sra_test_covmat))
	sra_test_covmat= sra_test_covmat[, -colremove]

	  sra_test_inputdata=list()
	  ov <- findOverlaps(predictor$regiondata,merge_input$regiondata)
	  index_regions <- subjectHits(ov)
	  #reorder GRanges objects
	  sra_test_inputdata$covmat  <- sra_test_covmat[index_regions,]
	  sra_test_inputdata$regiondata <- merge_input$regiondata[index_regions]

	save(sra_test_inputdata, file= "/dcl01/leek/data/sellis/barcoding/data/sra_test_inputdata_SequencingStrategy_nocovars.rda")
}else{
	load('/dcl01/leek/data/sellis/barcoding/data/sra_test_inputdata_SequencingStrategy_nocovars.rda')
}
```

# Predict Sequencing Strategy in SRA test set

```{r 'predict-phenotype', message = FALSE, warning = FALSE}
predictions<-predict_pheno(inputdata_test= sra_test_inputdata, phenodata=sra_meta_training, phenotype=phenot, covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions)

#since we know the truth here, let's check and see how we're doing:
 
  actual = sra_meta_test[,phenot]

  #define predicted
  predicted = predictions

  #summarize data
	number_match <- sum(predicted==actual)
	perc_correct = sum(predicted==actual)/length(actual)
	number_sites = nrow(predictor$coefEsts)

	summarized = cbind(number_sites,number_match, perc_correct)
	colnames(summarized) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 summarized


#get output for package
predictions_SequencingStrategy_test = as.data.frame(cbind(sra_meta_test$run,"sra",actual=as.character(actual),predicted,perc_correct))
#update training set with test set perc_correct
predictions_SequencingStrategy$perc_correct <- perc_correct

## how do predictions fare when sra_misreported_paired_end==TRUE
set <- actual!=predicted
mismatch = sra_meta_test$sra_misreported_paired_end[set]
table(mismatch)
#Proportion of samples with misreported paired end sequencing that are 'incorrect' predictions in SRA training data
length(mismatch[mismatch==TRUE])/length(sra_meta_test$sra_misreported_paired_end[sra_meta_test$sra_misreported_paired_end==TRUE])
```

# Predict Sequnencing Strategy in TCGA
```{r 'predict-phenotype-tcga', message = FALSE, warning = FALSE}

regions = predictor$regiondata

## save as bedfile for region extraction in TCGA
#compare to previous sex prediction
# gr = predictor$regiondata
# export.bed(gr, con="/dcl01/leek/data/sellis/barcoding/TCGA/TCGA_SequencingStrategy_v2.bed")

## run scripts/run_bwtool_SequencingStrategy.sh to extract regions
# system('sh /dcl01/leek/data/sellis/barcoding/scripts/run_bwtool_SequencingStrategy.sh tcga')
# data are here: /dcl01/leek/data/sellis/barcoding/TCGA/bwtool_SequencingStrategy/coverage_tcga_SequencingStrategy

## run check_tsv_tcga to check file and merge
# qrsh -l mem_free=2G,h_vmem=3G -pe local 25
# module load R/3.3.x
# R CMD '/dcl01/leek/data/sellis/barcoding/scripts/check_tsv_tcga.R'

load("/dcl01/leek/data/sellis/barcoding/data/regions_LibraryLayout_TCGA.rda")
covmat = cov_tcga

# get phenotype info in order
recount::all_metadata('TCGA') -> md 
bws = gsub('.bw','', md$bigwig_file)


#remove coverage samples for which we don't have metadata
covmat = covmat[,colnames(covmat) %in% bws]
## reorder
covmat = covmat[,bws]
#put data on a log2 scale
dm = log2(covmat+1)


tcga_test_inputdata = list()
tcga_test_inputdata$covmat <- dm
tcga_test_inputdata$regiondata <- regions

predictions_tcga<-predict_pheno(inputdata_test= tcga_test_inputdata, phenodata=sra_meta_training, phenotype=phenot, covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions_tcga)

#since we know the truth here, let's check and see how we're doing:
  actual = rep("PAIRED",length(predictions_tcga))

  #define predicted
  predicted = predictions_tcga

  #summarize data
	number_match <- sum(predicted==actual)
	perc_correct = sum(predicted==actual)/length(actual)
	number_sites = nrow(predictor$coefEsts)

	summarized_tcga = cbind(number_sites,number_match, perc_correct)
	colnames(summarized_tcga) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 summarized_tcga

 #get output for package
predictions_tcga = as.data.frame(cbind(md$gdc_file_id,"tcga",actual,predicted,perc_correct))
```

# Predict Library Type in GTEx

```{r 'predict-phenotype-gtex', message = FALSE, warning = FALSE}
#get GTEx metadata
load("/dcl01/leek/data/sellis/barcoding/data/sample_individuals.Rdata")
load("/dcl01/leek/data/sellis/barcoding/data/rda/gtexmetadata.rda")
gtex_meta = gtexmetadata
gtex_meta = cbind(gtex_meta,usegtex)
rm(gtexmetadata,usegtex)
gtex_meta[is.na(gtex_meta[,"SMTS"]),"SMTS"] <- c("Esophagus","Skin","Stomach","Skin","Esophagus")
usegtex = gtex_meta$usegtex
pheno = gtex_meta
pheno = pheno[usegtex,]

load("/dcl01/leek/data/sellis/barcoding/data/regions_LibraryLayout_GTEx.rda")
covmat = cov_gtex[,4:ncol(cov_gtex)]
covmat = covmat[,usegtex]

gtex_test_inputdata = list()
gtex_test_inputdata$covmat <- log2(covmat+1)
gtex_test_inputdata$regiondata <- regions

predictions_gtex<-predict_pheno(inputdata_test= gtex_test_inputdata, phenodata=sra_meta_training, phenotype=phenot, covariates=NULL,type="factor", predictordata = predictor)
# number of samples
length(predictions_gtex)

#since we know the truth here, let's check and see how we're doing:
  actual = rep("PAIRED",length(predictions_gtex))

  #define predicted
  predicted = predictions_gtex

  #summarize data
	number_match <- sum(predicted==actual)
	perc_correct = sum(predicted==actual)/length(actual)
	number_sites = nrow(predictor$coefEsts)

	summarized_gtex = cbind(number_sites,number_match, perc_correct)
	colnames(summarized_gtex) <- c("sites_tested", "number_correct", "percent_correct")
 
 #compare predictions to known sex
 summarized_gtex

 #get output for package
predictions_gtex = as.data.frame(cbind(pheno$Run,"gtex",actual,predicted,perc_correct))
```

```{r 'plot-output', message = FALSE, warning = FALSE}

data = rbind(summarized_training,summarized,summarized_gtex,summarized_tcga) %>% data.frame()
t(data) %>% as.matrix() -> mat
colnames(mat) <- c("SRA: training","SRA: test","GTEx", "TCGA")
mat 

colours2 = c(bright["teal"],bright["teal"],bright["purple"],bright["pink"])

# library(extrafont)
# pdf("/dcl01/leek/data/sellis/barcoding/plots/SequencingStrategyPrediction.pdf",family="Roboto Condensed",width=12, height=6)
par(cex=1.3)
barplot(as.numeric(mat["percent_correct",])*100, names.arg=c("SRA: training","SRA: test","GTEx","TCGA"),
        main="Sequencing Strategy Prediction", ylim=c(0,100), ylab="Accuracy", xlab="Data Set Used", col=colours2, cex.lab=1.3, cex.axis=1.2, cex.main=1.4)
axis(1, at=c(0,5), xpd=T, lwd.ticks = 0, labels=F)
# dev.off()

```

```{r 'pheno-output', message = FALSE, warning = FALSE}
predictions_SequencingStrategy <- rbind(predictions_SequencingStrategy,predictions_SequencingStrategy_test,predictions_gtex,predictions_tcga)
colnames(predictions_SequencingStrategy) <- c("sample_id","dataset","reported_sequencingstrategy","predicted_sequencingstrategy","accuracy_sequencingstrategy") 

predictions_SequencingStrategy$dataset <- factor(predictions_SequencingStrategy$dataset)
predictions_SequencingStrategy$reported_sequencingstrategy <- factor(predictions_SequencingStrategy$reported_sequencingstrategy)
predictions_SequencingStrategy$predicted_sequencingstrategy <- factor(predictions_SequencingStrategy$predicted_sequencingstrategy)
predictions_SequencingStrategy$accuracy_sequencingstrategy <- as.numeric(predictions_SequencingStrategy$accuracy_sequencingstrategy)

save(predictions_SequencingStrategy, file="/dcl01/leek/data/sellis/barcoding/output/predictions_SequencingStrategy.rda")
```


```{r 'region-information', message = FALSE, warning = FALSE}
## Get required information for the plots
if(!file.exists("/dcl01/leek/data/sellis/barcoding/output/SequencingStrategy_regions.rda")) {
	txdb <- GenomicFeatures::makeTxDbFromGFF('ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/gencode.v25.annotation.gff3.gz', format = 'gff3', organism = 'Homo sapiens')
	tx<-annotateTranscripts(txdb, annotationPackage = NULL, by = c("tx","gene"), codingOnly=FALSE, verbose = TRUE, requireAnnotation = FALSE)
	output<-matchGenes(predictor$regiondata, tx, type = c("any", "fiveprime"), promoterDist = 2500, skipExons = FALSE, verbose = TRUE)
	output
	save(output, file= "/dcl01/leek/data/sellis/barcoding/output/SequencingStrategy_regions.rda")
}else{
	load("/dcl01/leek/data/sellis/barcoding/output/SequencingStrategy_regions.rda")
	table(output$description)
}
```

# Vignette information

```{r reproducibility}
## Time spent creating this report:
diff(c(startTime, Sys.time()))

## Date this report was generated
message(Sys.time())

## Reproducibility info
options(width = 120)
devtools::session_info()
```

Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library('rmarkdown')
system.time(render('/dcl01/leek/data/sellis/barcoding/phenotype_vignettes/phenopredict-SequencingStrategy.Rmd', 'BiocStyle::html_document'))

## Extract the R code
library('knitr')
knit('/users/sellis/phenopredict/vignettes/phenopredict-SequencingStrategy.Rmd', tangle = TRUE)
```


